### PR TITLE
webpack v4 plugin authoring typedefs

### PR DESCRIPTION
**Pull request by dennispg**
Please fill in this template.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/webpack/webpack
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This is my first pass attempt at trying to implement definitions for all the new hooks available to plugin authors in v4.

There is still no clear documentation available anywhere, so I am scouring the source code to glean as much type information as I can from there. I have put `any` in several places where I am not sure what to use otherwise just yet...

Specifically for each of the new hooks available in `webpack.Compiler` and `webpack.Compilation` and others, each hook inherits from a subclass of `Tapable.Hook<T>`. `T` is meant to be used to strongly type the `context` that is passed into the callback provided to the `tap` method. I have typed the few hooks where I was able to track down plugins that use it and infer what type it should be.

Appreciate any guidance or contributions.

### CODE REVIEW

File: types/tapable/index.d.ts
Comment (armanio123) [inline types/tapable/index.d.ts]: Mentioning @e-cloud for review.

File: types/webpack/index.d.ts
Comment (armanio123) [inline types/webpack/index.d.ts]: Mentioning definition owners for review: @tkqubo @bumbleblym @bcherny @tommytroylin @mohsen1 @jcreamer898 @ahmed-taj @alan-agius4 @elliottsj @jason0x43

File: types/tapable/index.d.ts
Comment (mohsen1) [inline types/tapable/index.d.ts]: why you need to default to `any`?

File: types/webpack/index.d.ts
Comment (mohsen1) [inline types/webpack/index.d.ts]: string?

File: types/tapable/index.d.ts
Comment (dennispg) [reply types/tapable/index.d.ts]: I'm definitely looking for suggestions here because I think I might have gotten in over my head a bit, but here's my line of reasoning on this:

Take this set of hooks as a sample: https://github.com/webpack/webpack/blob/8a59ef79b3124f76cbc6b19f19b2276f06178b6d/lib/Compiler.js#L29
```javascript
shouldEmit: new SyncBailHook(["compilation"]),
done: new AsyncSeriesHook(["stats"]),
additionalPass: new AsyncSeriesHook([]),
beforeRun: new AsyncSeriesHook(["compilation"]),
run: new AsyncSeriesHook(["compilation"]),
emit: new AsyncSeriesHook(["compilation"]),
afterEmit: new AsyncSeriesHook(["compilation"]),
thisCompilation: new SyncHook(["compilation", "params"]),
compilation: new SyncHook(["compilation", "params"]),
normalModuleFactory: new SyncHook(["normalModuleFactory"]),
contextModuleFactory: new SyncHook(["contextModulefactory"]),
beforeCompile: new AsyncSeriesHook(["params"]),
compile: new SyncHook(["params"]),
make: new AsyncParallelHook(["compilation"]),
afterCompile: new AsyncSeriesHook(["compilation"]),
watchRun: new AsyncSeriesHook(["compiler"]),
failed: new SyncHook(["error"]),
invalid: new SyncHook(["filename", "changeTime"]),
watchClose: new SyncHook([]),
```
You can see that each hook accepts a different number of parameters... (The max I've seen in the source has been 3, I figured I would start with that many..

I wanted a way to strongly type the callback fns for these hooks, but from what I've been reading theres no way to do this right now.
 (https://github.com/Microsoft/TypeScript/issues/5453)

By defining Hook as I have, I was able to define these hooks like so:

```typescript
additionalPass: AsyncSeriesHook;
beforeRun: AsyncSeriesHook<Compilation>;
run: AsyncSeriesHook<Compilation>;
emit: AsyncSeriesHook<Compilation>;
afterEmit: AsyncSeriesHook<Compilation>;
thisCompilation: SyncHook<Compilation, { normalModuleFactory: NormalModuleFactory }>;
compilation: SyncHook<Compilation, { normalModuleFactory: NormalModuleFactory }>;
normalModuleFactory: SyncHook<NormalModuleFactory>;
```

And then when I tap the hook, the parameters are strongly typed. In this example:
```
compiler.hooks.compilation.tap(
    "DllEntryPlugin",
    (compilation, { normalModuleFactory }) => {
        const dllModuleFactory = new DllModuleFactory();
        compilation.dependencyFactories.set(
            DllEntryDependency,
            dllModuleFactory
        );
    }
);
```

The callback signature is equivalent to:
```typescript
(compilation: webpack.compilation.Compilation, { normalModuleFactory: webpack.compilation.NormalModuleFactory })
```

I really hope that makes sense... I'm kinda struggling with this.

File: types/webpack/index.d.ts
Comment (dennispg) [reply types/webpack/index.d.ts]: I've been using `any` in a few places like this as a stop gap till I am able to figure out what all the types are...

File: types/tapable/index.d.ts
Comment (mohsen1) [reply types/tapable/index.d.ts]: thanks for explaining. Makes sense

Overall I feel Webpack source is a good candidate for TypeScriptification. Not sure how open @sokra and @TheLarkInn are to this idea! I love to help type the code.

File: types/webpack/index.d.ts
Comment (tommytroylin) [inline types/webpack/index.d.ts]: I really appreciate this types for hooks.

File: types/tapable/index.d.ts
Comment (dennispg) [reply types/tapable/index.d.ts]: That would be the most ideal thing!

**Comment by armanio123** [comment]
Did we got an approval for this or still waiting for review?

**Review by mohsen1 — APPROVED** [review]

**Pull request merged**
